### PR TITLE
Configuration: Split options into global and per-module configuration

### PR DESCRIPTION
**Pull request by RAOF**
This lets each module define its own options, without needing the option names to be unique across all modules.

Also organises the module-specific options into their own sections at the end of the help text.

### CODE REVIEW

File: include/platform/mir/options/default_configuration.h [hunk 1]
```diff
@@ -74,6 +75,47 @@ class DefaultConfiguration : public Configuration
     char const** const argv;
     std::function<void(int argc, char const* const* argv)> const unparsed_arguments_handler;
     std::shared_ptr<boost::program_options::options_description> const program_options;
+
+    template<typename Value>
+    class LibraryMap
```
Comment (tarek-y-ismail) [inline include/platform/mir/options/default_configuration.h:80]: Why not use a `map` or `unordered_map`? Can't quite figure out the difference in functionality

**Comment by RAOF** [comment]
Aaargh. When did those tests break?

File: include/platform/mir/options/default_configuration.h (hunk 1 above)
Comment (RAOF) [reply include/platform/mir/options/default_configuration.h:80]: In the current code? It's got a slightly more pleasant interface for what we need.

As to why this exists? This is somewhat of a path-dependent thing. So, it was originally trying to be `unordered_map<SharedLibrary, ...>`, but `SharedLibrary` is non-copyable. So, I wanted `unordered_map<SharedLibrary*, ...>` with a custom equals and hash, but that was surprisingly difficult.

Now that we've gone to `SharedLibrary::Handle`, we could *probably* move back to a `std::unordered_map<SharedLibrary::Handle, ...>` with a slightly less pleasant interface (needing to call `.get_handle()` in various places).

File: src/server/graphics/default_configuration.cpp [hunk 2]
```diff
@@ -296,13 +296,15 @@ auto mir::DefaultServerConfiguration::the_rendering_platforms() ->
                         device.device->devnode());
                 }
 
+                mir::log_info("Hello? Do we see this for %s", description->name);
+
```
Comment (AlanGriffiths) [inline src/server/graphics/default_configuration.cpp:300]: Stray debug code?

File: src/server/graphics/platform_probe.cpp [hunk 3]
```diff
@@ -14,6 +14,7 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include "mir/graphics/default_display_configuration_policy.h"
```
Comment (AlanGriffiths) [inline src/server/graphics/platform_probe.cpp:17]: I guess it is pre-existing. But we try to ensure that the first header included is the one that corresponds. (In this case `#include "platform_probe.h"`

**Review by tarek-y-ismail — APPROVED** [review]

**Comment by RAOF** [comment]
Now with fewer conflicts :woman_facepalming:

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
Generally, I like the approach and the result.

In addition to the naming tweaks and mircommon stanza, this breaks ABI for mirplatform and mirserver

File: include/platform/mir/options/configuration.h [hunk 4]
```diff
@@ -66,17 +68,24 @@ extern char const* const auto_console;
...
 public:
-    virtual std::shared_ptr<options::Option> the_options() const = 0;
+    /**
+     * The options not associated with a specific loaded module
+     */
+    virtual std::shared_ptr<options::Option> global_options() const = 0;
+    /**
+     * All options, including those added by the specified module
+     */
+    virtual auto the_options_for(SharedLibrary const& module) const -> std::shared_ptr<options::Option> = 0;
```
[the remark this answers is not in the document]
Comment (RAOF) [reply include/platform/mir/options/configuration.h:81]: :+1: on `options_for(...)`

File: src/common/symbols.map [hunk 5]
```diff
@@ -691,8 +696,12 @@ global:
...
+global:
+  extern "C++" {
     MirTouchpadConfig::disable_with_external_mouse*;
     mir::event_type_to_c_str*;
+    std::hash::operator*;
+    typeinfo?for?std::hash;
   };
-} MIR_COMMON_2.17;
\ No newline at end of file
+} MIR_COMMON_2.17;
```
[the remark this answers is not in the document]
Comment (tarek-y-ismail) [reply src/common/symbols.map:706]: I think this is more of a symbol generator thing. I checked the code and here:
https://github.com/canonical/mir/blob/4d066de260f87df5162713b8343fe21835eab5e1/tools/symbols_map_generator/main.py#L544-L557

It uses the previous version when closing the stenza. I'm not very familiar with symbol files, so I'm not  sure if this is a bug or a misunderstanding.

File: src/common/symbols.map [hunk 6]
```diff
@@ -691,8 +696,12 @@ global:
 };
 
 MIR_COMMON_2.18 {
-  global: extern "C++" {
+global:
+  extern "C++" {
     MirTouchpadConfig::disable_with_external_mouse*;
     mir::event_type_to_c_str*;
+    std::hash::operator*;
+    typeinfo?for?std::hash;
```
Comment (AlanGriffiths) [inline src/common/symbols.map:704]: We're only providing a `hash` specialization on `mir::SharedLibrary::Handle`, we shouldn't be exporting arbitrary other instantiations (they are hard to control/check for changes).

File: src/common/symbols.map (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply src/common/symbols.map]: > I'm not very familiar with symbol files, so I'm not sure if this is a bug or a misunderstanding.

https://canonical-mir.readthedocs-hosted.com/stable/reference/dso_versioning_guide/

If the tooling isn't doing the right thing, then it needs fixing (or at least documenting as an issue)

File: src/common/symbols.map [hunk 7]
```diff
@@ -323,7 +323,10 @@ global:
     mir::RecursiveReadWriteMutex::write_lock*;
     mir::RecursiveReadWriteMutex::write_unlock*;
     mir::SharedLibrary::?SharedLibrary*;
+    mir::SharedLibrary::Handle::?Handle*;
+    mir::SharedLibrary::Handle::operator*;
     mir::SharedLibrary::SharedLibrary*;
+    mir::SharedLibrary::get_handle*;
```
[the remark this answers is not in the document]
Comment (mattkae) [reply src/common/symbols.map:329]: @tarek-y-ismail My guess here is that the minor version in the `CMakeLists.txt` isn't yet bumped to 2.19, so we're adding to the previous stanza. The generator isn't smart enough to know when a release happened (although that would be pretty interesting to do...)

[the remark this answers is not in the document]
Comment (tarek-y-ismail) [reply src/common/symbols.map:329]: Maybe it'd be easier to just assume `<major version>.(<minor version> + 1)`?

File: src/common/symbols.map (hunk 5 above)
[the remark this answers is not in the document]
Comment (tarek-y-ismail) [reply src/common/symbols.map]: @mattkae The inclusion of `std::hash::operator*` should also be taken into account. Perhaps we shouldn't export specific symbols when it comes to `std`?

File: src/common/symbols.map (hunk 7 above)
[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply src/common/symbols.map:329]: > so we're adding to the previous stanza

Actually, these are wrongly added to the 2.17 stanza, not wrongly added to the 2.18 stanza.

Anyway, for **this PR** the thing that needs fixing is the symbols.map, any issue with the script can be filed as a bug

File: src/common/symbols.map [hunk 8]
```diff
@@ -485,6 +485,7 @@ global:
     non-virtual?thunk?to?mir::logging::Logger::log*;
     non-virtual?thunk?to?mir::logging::MultiLogger::log*;
     operator*;
+    std::hash?mir::SharedLibrary::Handle?::operator*;
```
Comment (AlanGriffiths) [inline src/common/symbols.map:488]: Does not belong to 2.18

File: src/common/symbols.map [hunk 9]
```diff
@@ -499,6 +500,7 @@ global:
     typeinfo?for?mir::IntOwnedFd;
     typeinfo?for?mir::NonBlockingExecutor;
     typeinfo?for?mir::PosixRWMutex;
+    typeinfo?for?mir::SharedLibrary::Handle;
```
Comment (AlanGriffiths) [inline src/common/symbols.map:503]: Does not belong to 2.18

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]

**Comment by AlanGriffiths** [comment]
Oops! The "Symbols Check" is failing because the `symbols_map_generator` isn't matching the corrected `symbols.map`...

```
INFO:__main__:symbols_map_generator is running in 'diff' mode
INFO:__main__:MIR_SYMBOLS_MAP_GENERATOR_CLANG_SO_PATH is /usr/lib/llvm-18/lib/libclang-18.so.1
INFO:__main__:MIR_SYMBOLS_MAP_GENERATOR_CLANG_LIBRARY_PATH is /usr/lib/llvm-18/lib
INFO:__main__:Symbols map generation is beginning for library=mir_common with version=2.18
INFO:__main__:Processing external headers directory: /home/runner/work/mir/mir/include/common
INFO:__main__:Processing internal headers directory: /home/runner/work/mir/mir/src/include/common
External Symbols Diff:

  New Symbols 🟢🟢🟢
    std::hash::operator*;
    typeinfo?for?std::hash;

  Deleted Symbols 🔻🔻🔻

Internal Symbols Diff:

gmake[3]: *** [src/common/CMakeFiles/check-mircommon-symbols-map.dir/build.make:70: src/common/CMakeFiles/check-mircommon-symbols-map] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:3151: src/common/CMakeFiles/check-mircommon-symbols-map.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:3158: src/common/CMakeFiles/check-mircommon-symbols-map.dir/rule] Error 2
gmake: *** [Makefile:621: check-mircommon-symbols-map] Error 2
Error: Process completed with exit code 2.
```

...I guess it does need fixing after all

**Comment by tarek-y-ismail** [comment]
> Oops! The "Symbols Check" is failing because the symbols_map_generator isn't matching the corrected symbols.map...

Hmm, that's weird. It seems that the code builds fine without them. I  just thought I accidentally added them so I deleted them...

**Comment by AlanGriffiths** [comment]
> Hmm, that's weird. It seems that the code builds fine without them

Nothing weird about it. The Check runs `--target check-mircommon-symbols-map`, which isn't part of `all`.

The problem is that the script wrongly expects the `std::hash` symbols

**Comment by AlanGriffiths** [comment]
Maybe the way forward is to inline the `std::hash` specialization (if it is inline it shouldn't be exporting symbols)

```C++
template<>
struct std::hash<mir::SharedLibrary::Handle>
{
    auto operator()(mir::SharedLibrary::Handle const& handle) const noexcept -> std::size_t
    {
// either
        return std::hash<void*>{}(handle.handle);
// or 
        return handle.hash(); // With a hash() implementation in `SharedLibrary::Handle`
    }
};
```

**Comment by AlanGriffiths** [comment]
@tarek-y-ismail you should be able to test the symbols check with

    MIR_SYMBOLS_MAP_GENERATOR_CLANG_SO_PATH=/usr/lib/llvm-18/lib/libclang-18.so.1 MIR_SYMBOLS_MAP_GENERATOR_CLANG_LIBRARY_PATH=/usr/lib/llvm-18/lib\
    ninja -C build  check-mircommon-symbols-map

**Comment by tarek-y-ismail** [comment]
> you should be able to test the symbols check with

As I expected, the generator script still exports `std::hash` and  the checking script still complains about it. Even though the project builds fine.
```
External Symbols Diff:

  New Symbols 🟢🟢🟢
    std::hash::operator*;
    typeinfo?for?std::hash;

  Deleted Symbols 🔻🔻🔻

Internal Symbols Diff:
```

**Comment by AlanGriffiths** [comment]
> As I expected, the generator script still exports `std::hash` and the checking script still complains about it. Even though the project builds fine.

yes, the script is doing the wrong thing. Which wouldn't matter if we didn't rely on it for symbols checking

**Review by AlanGriffiths — APPROVED** [review]

**Pull request merged**